lib/compat/wordpress-6.1/block-template-utils.php

```diff
@@ -290,3 +290,43 @@ function gutenberg_build_block_template_result_from_post( $post ) {
 	}
 	return $template;
 }
+
+/**
+ * Helper function to get the Template Hierarchy for a given slug.
+ * We need to Handle special cases here like `front-page`, `singular` and `archive` templates.
+ *
+ * Noting that we always add `index` as the last fallback template.
+ *
+ * @param string $slug The slug from which to extract the template hierarchy.
+ * @return array<string> The template hierarchy.
+ */
+function gutenberg_get_template__hierarchy( $slug ) {
+	if ( 'front-page' === $slug ) {
+		return array( 'front-page', 'home', 'index' );
+	}
```

Comment:
If possible, I'd rather handle special cases like this differently by mimicking what the frontend template resolution algorithm does.

There are two "levels" to this:
1. The "micro" level: Falling back within a given template type, e.g. from `single-post-hello-world` to `single-post` to `single`.
2. The "macro" level: Falling back from one template type to another (and finally to `index`). `front-page` to `home` to `index` falls into this category.

On the frontend, the "macro" level handling is done [here](https://github.com/WordPress/wordpress-develop/blob/9db156e7c73774836e3da71d8aac8c0adc2cf5e9/src/wp-includes/template-loader.php#L57-L95), basically by iterating over template types in a certain order, and checking which of them applies (that's the `is_${template_type}()` checks).

The "micro" level is then handled by the `get_${template_type}_template()` functions (e.g. [`get_single_template()`](https://github.com/WordPress/wordpress-develop/blob/9db156e7c73774836e3da71d8aac8c0adc2cf5e9/src/wp-includes/template.php#L514-L568)). 

While we cannot directly re-use either part (since they're tightly coupled to the frontend, the current route, and the `WP_Query` object), we can at least mimic their structure (and eventually maybe share _some_ code).

Keeping the "macro" and "micro" handling separate would specifically mean that we don't call `resolve_block_template` with a `$template_hierarchy` array that contains fallbacks of different template types (e.g. `array( 'front-page', 'home', 'index' )`) but that we iterate over the different template types until we find a match, and only pass a `$template_hierarchy` that contains template slugs for the current template type (e.g. `array( 'single-post-hello-world', 'single-post', 'single' )`.

I'll add more comments to better explain what I mean in the relevant spots.

Keeping things as similar as possible to frontend template resolution should make it easier to reason about, and much less error-prone.

Reply:
Thanks for the feedback Bernie! I'll have to study more the internals of core handling to process these info, but for start I'd like to know if you saw something that doesn't work as it should right now.

My initial thoughts(that might be wrong as I need to study more the core implementation to comprehend them 😄) are:
1. I mainly rely on [resolve_block_template](https://developer.wordpress.org/reference/functions/resolve_block_template/) and its logic, so if we wanted to mimic more of the internal core implementation, should we change that function?
2. I can't see right now why we would need to mimic some internal implementation and also differentiate the micro/macro types. For me without knowing all the implementation details I could make this work by just following the [template hierarchy diagram](https://developer.wordpress.org/files/2014/10/Screenshot-2019-01-23-00.20.04.png) - assuming it's working properly 😄 .
3. We are going to have some special handling such as `404` where we'll explicitly provide some fallback content and this will deviate from the front-end resolution, which  I don't think that's a bad thing.

Reply:
I'll mark this resolved - see: https://github.com/WordPress/gutenberg/pull/42520#issuecomment-1195688285

We can start new discussions since this code is outdated now.